src/algebra/group_with_zero.lean

```diff
@@ -0,0 +1,504 @@
+  classical,
+  by_cases hx : x = 0, { simp [hx] },
+  by_cases hy : y = 0, { simp [hy] },
+  symmetry,
+  apply eq_inv_of_mul_left_eq_one,
+  simp [mul_assoc, hx, hy]
+end
+
+theorem inv_comm_of_comm {a b : G} (H : a * b = b * a) : a⁻¹ * b = b * a⁻¹ :=
+begin
+  have : a⁻¹ * (b * a) * a⁻¹ = a⁻¹ * (a * b) * a⁻¹ :=
+    congr_arg (λ x:G, a⁻¹ * x * a⁻¹) H.symm,
+  classical,
+  by_cases h : a = 0, { rw [h, inv_zero, zero_mul, mul_zero] },
+  rwa [inv_mul_cancel_assoc_right _ _ h, mul_assoc, mul_inv_cancel_assoc_left _ _ h] at this,
+end
+
+@[simp] lemma div_self {a : G} (h : a ≠ 0) : a / a = 1 := mul_inv_cancel _ h
+
+@[simp] lemma div_one (a : G) : a / 1 = a :=
+show a * 1⁻¹ = a, by rw [inv_one, mul_one]
+
+lemma one_div (a : G) : 1 / a = a⁻¹ := one_mul _
+
+@[simp] lemma zero_div (a : G) : 0 / a = 0 := zero_mul _
+
+@[simp] lemma div_zero (a : G) : a / 0 = 0 :=
+show a * 0⁻¹ = 0, by rw [inv_zero, mul_zero]
+
+@[simp] lemma div_mul_cancel (a : G) {b : G} (h : b ≠ 0) : a / b * b = a :=
+inv_mul_cancel_assoc_left a b h
+
+@[simp] lemma mul_div_cancel (a : G) {b : G} (h : b ≠ 0) : a * b / b = a :=
+mul_inv_cancel_assoc_left a b h
+
+lemma mul_div_assoc {a b c : G} : a * b / c = a * (b / c) :=
+mul_assoc _ _ _
+
+local attribute [simp]
+div_eq_mul_inv mul_comm mul_assoc
+mul_left_comm mul_inv_cancel inv_mul_cancel
+
+lemma div_eq_mul_one_div (a b : G) : a / b = a * (1 / b) :=
+by simp
+
+lemma mul_one_div_cancel {a : G} (h : a ≠ 0) : a * (1 / a) = 1 :=
+by simp [h]
+
+lemma one_div_mul_cancel {a : G} (h : a ≠ 0) : (1 / a) * a = 1 :=
+by simp [h]
+
+lemma one_div_one : 1 / 1 = (1:G) :=
+div_self (ne.symm zero_ne_one)
+
+lemma one_div_ne_zero {a : G} (h : a ≠ 0) : 1 / a ≠ 0 :=
+assume : 1 / a = 0,
+have 0 = (1:G), from eq.symm (by rw [← mul_one_div_cancel h, this, mul_zero]),
+absurd this zero_ne_one
+
+lemma mul_ne_zero_comm {a b : G} (h : a * b ≠ 0) : b * a ≠ 0 :=
+by { rw mul_ne_zero_iff at h ⊢, rwa and_comm }
+
+lemma eq_one_div_of_mul_eq_one {a b : G} (h : a * b = 1) : b = 1 / a :=
+have a ≠ 0, from
+   assume : a = 0,
+   have 0 = (1:G), by rwa [this, zero_mul] at h,
+      absurd this zero_ne_one,
+have b = (1 / a) * a * b, by rw [one_div_mul_cancel this, one_mul],
+show b = 1 / a, by rwa [mul_assoc, h, mul_one] at this
+
+lemma eq_one_div_of_mul_eq_one_left {a b : G} (h : b * a = 1) : b = 1 / a :=
+have a ≠ 0, from
+  assume : a = 0,
+  have 0 = (1:G), by rwa [this, mul_zero] at h,
+    absurd this zero_ne_one,
+by rw [← h, mul_div_assoc, div_self this, mul_one]
+
+@[simp] lemma one_div_div (a b : G) : 1 / (a / b) = b / a :=
+by rw [one_div, div_eq_mul_inv, mul_inv_rev, inv_inv, div_eq_mul_inv]
+
+@[simp] lemma one_div_one_div (a : G) : 1 / (1 / a) = a :=
+by simp
+
+lemma eq_of_one_div_eq_one_div {a b : G} (h : 1 / a = 1 / b) : a = b :=
+by rw [← one_div_one_div a, h, one_div_one_div]
+
+end gwz
+
+section not_comm
+
+variables {G : Type*} [group_with_zero G]
+variables {a b c : G}
+
+@[simp] lemma inv_eq_zero {a : G} : a⁻¹ = 0 ↔ a = 0 :=
```

Comment:
This should follow immediately from inv_involutive and inv_zero by some machine, which says that if you have any bijection then x=y iff f(x)=f(y); you can avoid the case split this way.

Reply:
Golfed. Thanks.